lib/checkassert.cpp

```diff
@@ -53,45 +54,103 @@ void CheckAssert::assertWithSideEffects()
-                continue;
+            if (isVariableAssignment(tmp)) {
+                const auto *var = tmp->astOperand1()->variable();
+                if (checkVariableAssignment(var, tok->scope())) {
+                    assignmentInAssertError(tmp, var->name());
+                }
+            }
 
-            const Function* f = tmp->function();
-            if (f->nestedIn->isClassOrStruct() && !f->isStatic() && !f->isConst()) {
-                sideEffectInAssertError(tmp, f->name()); // Non-const member function called
+            if (tmp->tokType() != Token::eFunction)   // TODO: constructor calls with initializer list are not detected!
                 continue;
+            const Function *func = tmp->function();
+            auto args = getArguments(tmp);
+            if (isFunctionWithSideEffect(func, {&args, tok->scope()})) {
+                sideEffectInAssertError(tmp, func->name());
             }
-            const Scope* scope = f->functionScope;
-            if (!scope) continue;
-
-            for (const Token *tok2 = scope->bodyStart; tok2 != scope->bodyEnd; tok2 = tok2->next()) {
-                if (!tok2->isAssignmentOp() && tok2->tokType() != Token::eIncDecOp)
-                    continue;
-
-                const Variable* var = tok2->previous()->variable();
-                if (!var || var->isLocal() || (var->isArgument() && !var->isReference() && !var->isPointer()))
-                    continue; // See ticket #4937. Assigning function arguments not passed by reference is ok.
-                if (var->isArgument() && var->isPointer() && tok2->strAt(-2) != "*")
-                    continue; // Pointers need to be dereferenced, otherwise there is no error
-
-                bool noReturnInScope = true;
-                for (const Token *rt = scope->bodyStart; rt != scope->bodyEnd; rt = rt->next()) {
-                    if (rt->str() != "return") continue; // find all return statements
-                    if (inSameScope(rt, tok2)) {
-                        noReturnInScope = false;
-                        break;
+        }
+        tok = endTok;
+    }
+}
+
+bool CheckAssert::isFunctionWithSideEffect(const Function *function, argumentCheck argsChecking) const {
+    if (function->isConstexpr()) return false;
+    if (function->nestedIn->isClassOrStruct() && !function->isConstructor()) {
+        if (!(function->isConst() || function->isStatic())) {
+            return true;
+        }
+    }
+    const Scope *scope = function->functionScope;
+    if (!scope) return false;
+
+    if (function->isConstructor()) {
+        const auto *initializationList =
+            function->constructorMemberInitialization();
+        if (initializationList) {
+            for (const Token *tok = initializationList; tok != scope->bodyStart; tok = tok->next()) {
+                if (tok->tokType() == Token::eIncDecOp) {
+                    const Variable *var = tok->previous()->variable();
```

Comment:
you probably copy/pasted this.. but it seems wrong to use `tok->previous()`.. what do you think?